tower-http/src/limit.rs

```diff
@@ -0,0 +1,246 @@
+    inner: S,
+    limit: usize,
+    _ty: PhantomData<fn() -> B>,
+}
+
+impl<S, B> LengthLimited<S, B> {
+    define_inner_service_accessors!();
+
+    /// Create a new `LengthLimited` with the given body length limit.
+    pub fn new(inner: S, limit: usize) -> Self {
+        Self {
+            inner,
+            limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<S, B> Clone for LengthLimited<S, B>
+where
+    S: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            limit: self.limit,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<S, B> Copy for LengthLimited<S, B> where S: Copy {}
+
+impl<S, B> fmt::Debug for LengthLimited<S, B>
+where
+    S: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("LengthLimited")
+            .field("inner", &self.inner)
+            .field("service", &format_args!("{}", any::type_name::<B>()))
+            .finish()
+    }
+}
+
+impl<ReqBody, ResBody, S> Service<Request<ReqBody>> for LengthLimited<S, ReqBody>
+where
+    S: Service<Request<Limited<ReqBody>>, Response = Response<ResBody>>,
+    S::Error: Into<BoxError>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+    ResBody::Error: Into<BoxError>,
+{
+    type Response = Response<UnsyncBoxBody<Bytes, BoxError>>;
+    type Error = BoxError;
+    type Future = ResponseFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(|e| e.into())
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let (parts, body) = req.into_parts();
+        let body = Limited::new(body, self.limit);
+        let req = Request::from_parts(parts, body);
+
+        ResponseFuture {
+            future: self.inner.call(req),
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`LengthLimit`].
+    pub struct ResponseFuture<F> {
+        #[pin]
+        future: F,
+    }
+}
+
+impl<F, ResBody, E> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    E: Into<BoxError>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+    ResBody::Error: Into<BoxError>,
+{
+    type Output = Result<Response<UnsyncBoxBody<Bytes, BoxError>>, BoxError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().future.poll(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Ok(data)) => {
+                let (parts, body) = data.into_parts();
+                let body = body.map_err(|err| err.into()).boxed_unsync();
+                let resp = Response::from_parts(parts, body);
+
+                Poll::Ready(Ok(resp))
+            }
+            Poll::Ready(Err(err)) => {
+                let err = err.into();
+                if let Some(_) = err.downcast_ref::<LengthLimitError>() {
```

Comment:
Hm does this actually work? This checks the error for `call`, not the `Body::poll_data` which is what yields `LengthLimitError`.

Reply:
It does, but there may be some restrictions that require a bit more to make this work as a layer. In particular, if the body is never read by the handler, then there will never be any error returned for the body being too large. For reference, note that the doctest for this module requires the handler to consume the body.

However, if there is an attempt to consume the body, and it is consumed beyond the given limit, the read of the body is what will trigger the error. We'd expect this error to be propagated out of the underlying service in some way.

Right now, this does expect that the top-level error be the (boxed) `LengthLimitError`, so if someone propagates the error by wrapping it inside an error enum, this wouldn't pick it out right. I can make a modification here to try to walk the source chain, which would then cover this case.

Reply:
I see. I don't think we should change the error to `BoxError`. That makes it harder to use with axum. I imagined someone would try and consume the body, if that fails they'd downcast the error to `LengthLimitError` and build a response themselves from that. Then the error type wouldn't have to change.

Reply:
Consumers wouldn't need to return `BoxError` themselves, but just an error type that can be converted into a `BoxError`. Let me put my thought together into this PR and we can take a look at it from there. I'll also update the doctest to show how a custom, non-`BoxError` would look.

Reply:
Also, if an underlying service wanted to trap the underlying `LengthLimitError` themselves, they would still be free to do that. This would just act as a backstop to attempt to return a correct response. Without it, hyper would just sever the connection on a service returning an error. See the comments in `hyper:service::http`.

Because Axum services always require that their handlers end up returning something that can be a `Response` and have an `Infallible` error type, this backstop may not be useful in the Axum context, but for those writing handlers directly on top of hyper, this could do the right thing for the difficult chunked body case.

Reply:
> Consumers wouldn't need to return BoxError themselves, but just an error type that can be converted into a BoxError

Yes and that will make it unergonomic to use with axum as it _forces_ the error type to be `BoxError` even if your service actually never fails. It'll require doing this https://docs.rs/axum/latest/axum/error_handling/index.html#applying-fallible-middleware. Therefore I still think we should keep the error type from the inner service.

Reply:
I don't think using `MapRequestBody` is a good workaround as you loose out on `content-length` checking.

Reply:
I wrote the axum integration and use case I have in mind [here](https://github.com/tokio-rs/axum/pull/1048).

Reply:
`MapRequestBody` as a suggestion isn't meant to be a workaround, but rather a pointer for users that want to impose a body limit _without_ also imposing a content-length check and automatic response and to head off any request to make this middleware configureable in that way.

Reply:
Did a couple example runs with the Axum integration you provided.

One byte too long, with immedate follow-on request (_hyper_ is able to resync safely):
![bildo](https://user-images.githubusercontent.com/18165/169891529-42660b3f-5e41-4e57-bd82-75d19f864911.png)

With a chunked body that is beyond _hyper_'s resync attempt. In this case _hyper_ severs the connection since it cannot resume (so second request is dropped):
```
HTTP/1.1 413 Payload Too Large
content-type: text/plain; charset=utf-8
content-length: 56
date: Mon, 23 May 2022 19:24:30 GMT

Failed to buffer the request body: length limit exceeded
```

And what the middleware returns when it rejects the request on `Content-Length` grounds (just a little over the limit, so _hyper_ resyncs):
```
HTTP/1.1 413 Payload Too Large
content-type: text/plain; charset=utf-8
content-length: 21
date: Mon, 23 May 2022 19:29:14 GMT

length limit exceeded
```

If the payload is significantly too large, then _hyper_ fails to resync after one attempt and severs the connection.

Reply:
What are the implications of that? That clients will have to open a new connection?

Reply:
If hyper can't re-sync, it will indeed sever the connection and the direct client would need to re-establish a connection. If the request is proxied by a load balancer of some sort, it may have better tolerances for re-syncing the incoming connection. This is server-dependent, however, so I point this out in the documentation without overly-alluding to _hyper_'s specific implementation.

Basically, hyper attempts to read one more chunk (with `transfer-encoding: chunked`) or one more buffer full (with `content-length: <len>`). If it hasn't reached the end of the current HTTP message frame by that point, then it will kill the connection to reset things to a known acceptable state.